File: problems/003-nth-fibonacci/ruby/optimized.rb
# Same O(n) window-slide as clean.rb; unrolled 2x so the loop body advances
# two Fibonacci steps per iteration-counter check -- mirrors the 8080
# optimized.s's peel-odd-n + unrolled-pair trick (see that file's comments
# for the full derivation), a genuine constant-factor win that stays in
# the manifest's declared O(n) class.
def solve(c)
  n = c["n"]
  a, b = 0, 1
  if n.odd?
    a, b = b, a + b
    n -= 1
  end
  while n > 0
    t = a + b
    b = t + b
    a = t
    n -= 2
  end
  a
end

File: problems/003-nth-fibonacci/javascript/optimized.js
// L1-fib-optimized: same O(n) window-slide as clean; unrolled 2x so the
// loop body advances two Fibonacci steps per iteration-counter check --
// mirrors the 8080 optimized.s's peel-odd-n + unrolled-pair trick, expressed
// in JS terms (a genuine constant-factor win, same complexity class --
// matches the manifest's declared O(n) for the optimized variant).
module.exports = function solve(c) {
  let a = 0, b = 1, i = c.n;
  if (i & 1) { const t = a + b; a = b; b = t; i--; }
  for (; i > 0; i -= 2) {
    const t = a + b;
    b = t + b;
    a = t;
  }
  return a;
};
